turtle_tf2_cpp/src/turtle_tf2_message_filter.cpp

```diff
@@ -0,0 +1,96 @@
+// Copyright 2021 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <rclcpp/rclcpp.hpp>
+#include <geometry_msgs/msg/point_stamped.hpp>
+
+#include <tf2_ros/transform_listener.h>
+#include <tf2_ros/message_filter.h>
+#include <tf2_ros/buffer.h>
+#include <tf2_ros/create_timer_ros.h>
+#ifdef TF2_CPP_HEADERS
+  #include <tf2_geometry_msgs/tf2_geometry_msgs.hpp>
+#else
+  #include <tf2_geometry_msgs/tf2_geometry_msgs.h>
+#endif
+#include <message_filters/subscriber.h>
+
+#include <chrono>
+#include <memory>
+#include <string>
+
+using std::placeholders::_1;
+using namespace std::chrono_literals;
+
+
+class PoseDrawer : public rclcpp::Node
+{
+public:
+  PoseDrawer()
+  : Node("turtle_tf2_pose_drawer")
+  {
+    auto node = rclcpp::Node::make_shared("tf2_ros_message_filter");
```

Comment:
I see you are creating a Node instance two times. Is there any particular reason for that? One called `"turtle_tf2_pose_drawer"` and one `"tf2_ros_message_filter"`. I see that in the code you are using `node->` and `this->` interchangeably, and this might be confusing to users. I recommend to stick with leaving the  `PoseDrawer() : Node("turtle_tf2_pose_drawer")` call and using this-> on all onwards, when you need to call your node

Reply:
@kurshakuz  You're correct. The reason I used them interchangeably is that when I create the messagefilter object,  tf2_filter_ = std::make_shared<tf2_ros::MessageFilter<geometry_msgs::msg::PointStamped>>(), the fifth argument need to be type ‘const SharedPtr&’ {aka ‘const std::shared_ptr<rclcpp::Node>&’}, while this pointer is type of ‘PoseDrawer*’, I don't know how to convert between them. Do you have any suggestion?

Reply:
Did you do it like below?
```
tf2_filter_ = std::make_shared<tf2_ros::MessageFilter<geometry_msgs::msg::PointStamped>>(
      point_sub_, *tf2_buffer_, target_frame_, 10, this, buffer_timeout);
```
I am not quite getting it as the PoseDrawer actually inherits from the Node type. Can you try to build it as above? If it is still does not work, try looking for other APIs of MessageFilter initialization. Some of them require node clock rather than node itself:
https://docs.ros2.org/foxy/api/tf2_ros/classtf2__ros_1_1MessageFilter.html

Let me know if it still does not work

Reply:
Yes, that does not work, when I build the turtle_tf2_cpp package, it will get errors:
from /home/wangjg/geometry_tutorials/turtle_tf2_cpp/src/turtle_tf2_message_filter.cpp:15:
/usr/include/c++/9/ext/new_allocator.h:145:20: note:   cannot convert ‘std::forward<PoseDrawer*>((* & __args#4))’ (type ‘PoseDrawer*’) to type ‘const SharedPtr&’ {aka ‘const std::shared_ptr<rclcpp::node_interfaces::NodeLoggingInterface>&’}
  145 |  noexcept(noexcept(::new((void *)__p)
      |                    ^~~~~~~~~~~~~~~~~~
  146 |        _Up(std::forward<_Args>(__args)...)))
      |        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /home/wangjg/geometry_tutorials/turtle_tf2_cpp/src/turtle_tf2_message_filter.cpp:21:
/home/wangjg/ros2_galactic_with_nav2/ros2_ws/install/tf2_ros/include/tf2_ros/message_filter.h:217:3: note: candidate: ‘template<class F, class TimeRepT, class TimeT> tf2_ros::MessageFilter<M, BufferT>::MessageFilter(F&, BufferT&, const string&, uint32_t, const SharedPtr&, std::chrono::duration<TimeRepT, TimeT>)’
  217 |   MessageFilter(
      |   ^~~~~~~~~~~~~

Reply:
I just dived into that and found that currently `tf2_ros::MessageFilter` is lacking that capabilities. There is an open issue on that here:
https://github.com/ros2/geometry2/issues/95

I suggest to leave it that way for now, as we will still be able to make corresponding updates later

Reply:
@kurshakuz  this problem was resolved just now:
tf2_filter_ = std::make_shared<tf2_ros::MessageFilter<geometry_msgs::msg::PointStamped>>(
      point_sub_, *tf2_buffer_, target_frame_, 10, this->get_node_logging_interface(),
      this->get_node_clock_interface(), buffer_timeout);
I used rclcpp::node_interfaces:NodeLoggingInterface and rclcpp::node_interfaces:NodeClockInterface other than  const rclcpp::Node::SharedPtr &, and it will work.